File: front-end/src/domain/valueobject/demand/daily-waste-demand-liters-per-day.ts
/**
 * DailyWasteDemandLitersPerDay
 *
 * Represents the approximate waste demand of a container per day in liters.
 * Immutable value object that encapsulates the daily demand value in liters per day.
 * The unit is always liters per day (fixed).
 */
export class DailyWasteDemandLitersPerDay {
  /**
   * Demand value in liters per day.
   * Must be non-negative.
   */
  private readonly litersPerDay: number;

  /**
   * Create a new `DailyWasteDemandLitersPerDay`.
   * @param litersPerDay The demand in liters per day (must be >= 0)
   * @throws Error if litersPerDay is negative
   */
  constructor(litersPerDay: number) {
    if (litersPerDay < 0) {
      throw new Error('Daily waste demand cannot be negative');
    }
    this.litersPerDay = litersPerDay;
  }

  /**
   * Get the demand value in liters per day.
   * @returns The demand in liters per day
   */
  getLitersPerDay(): number {
    return this.litersPerDay;
  }

  /**
   * Create a new instance with an updated value.
   * @param newLitersPerDay The new demand value in liters per day
   * @returns A new DailyWasteDemandLitersPerDay instance
   * @throws Error if newLitersPerDay is negative
   */
  setLitersPerDay(newLitersPerDay: number): DailyWasteDemandLitersPerDay {
    return new DailyWasteDemandLitersPerDay(newLitersPerDay);
  }

  /**
   * Add two DailyWasteDemandLitersPerDay instances.
   * @param other The other DailyWasteDemandLitersPerDay to add
   * @returns A new DailyWasteDemandLitersPerDay with the sum
   * @throws Error if other is null or undefined\n   
   */
   add(other: DailyWasteDemandLitersPerDay): DailyWasteDemandLitersPerDay {
    if (!other) {
      throw new Error('Other DailyWasteDemandLitersPerDay cannot be null');
    }
    return new DailyWasteDemandLitersPerDay(this.litersPerDay + other.litersPerDay);
  }

  /**
   * Compare if this demand is greater than another.
   * @param other The other DailyWasteDemandLitersPerDay to compare
   * @returns true if this demand is greater, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThan(other: DailyWasteDemandLitersPerDay): boolean {
    if (!other) {
      throw new Error('Other DailyWasteDemandLitersPerDay cannot be null');
    }
    return this.litersPerDay > other.litersPerDay;
  }

  /**
   * Compare if this demand is greater than or equal to another.
   * @param other The other DailyWasteDemandLitersPerDay to compare
   * @returns true if this demand is greater or equal, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThanOrEqual(other: DailyWasteDemandLitersPerDay): boolean {
    if (!other) {
      throw new Error('Other DailyWasteDemandLitersPerDay cannot be null');
    }
    return this.litersPerDay >= other.litersPerDay;
  }

  /**
   * Compare if this demand is less than another.
   * @param other The other DailyWasteDemandLitersPerDay to compare
   * @returns true if this demand is less, false otherwise
   * @throws Error if other is null or undefined
   */
  lessThan(other: DailyWasteDemandLitersPerDay): boolean {
    if (!other) {
      throw new Error('Other DailyWasteDemandLitersPerDay cannot be null');
    }
    return this.litersPerDay < other.litersPerDay;
  }

  /**
   * Compare equality with another object.
   * @param otherObject The object to compare with
   * @returns true if the objects are equal, false otherwise
   */
  equals(otherObject: unknown): boolean {
    if (this === otherObject) {
      return true;
    }
    if (!(otherObject instanceof DailyWasteDemandLitersPerDay)) {
      return false;
    }
    return this.litersPerDay === otherObject.litersPerDay;
  }

  /**
   * Get the hash code for this object.
   * @returns A hash code based on the litersPerDay value
   */
  hashCode(): number {
    return this.litersPerDay;
  }

  /**
   * Get a string representation of this value object.
   * @returns A formatted string representation
   */
  toString(): string {
    return `DailyWasteDemandLitersPerDay={litersPerDay=${this.litersPerDay}}`;
  }
}

File: back-end/src/main/java/es/ull/project/domain/valueobject/demand/DailyWasteDemandLitersPerDay.java
package es.ull.project.domain.valueobject.demand;

import java.util.Objects;

/**
 * DailyWasteDemandLitersPerDay
 *
 * Represents the approximate waste demand of a container per day in liters.
 * Immutable value object that encapsulates the daily demand value in liters per day.
 * The unit is always liters per day (fixed).
 */
public final class DailyWasteDemandLitersPerDay {

    /**
     * Error messages for validation.
     */
    private static final String ERROR_DEMAND_NEGATIVE = "Daily waste demand cannot be negative";
    private static final String ERROR_OTHER_DEMAND_NULL = "Other DailyWasteDemandLitersPerDay cannot be null";
    private static final int ZERO = 0;

    /**
     * Required.
     * Demand value in liters per day.
     * It is a required attribute and always measured in liters per day.
     */
    private final double litersPerDay;

    /**
     * Creates a new DailyWasteDemandLitersPerDay.
     *
     * @param litersPerDay Demand value in liters per day (must be ≥ 0)
     * @throws IllegalArgumentException if litersPerDay is negative
     */
    public DailyWasteDemandLitersPerDay(double litersPerDay) {
        if (litersPerDay < ZERO) {
            throw new IllegalArgumentException(ERROR_DEMAND_NEGATIVE);
        }
        this.litersPerDay = litersPerDay;
    }

    /**
     * Returns the demand value in liters per day.
     *
     * @return Demand value in liters per day
     */
    @com.fasterxml.jackson.annotation.JsonValue
    public double getLitersPerDay() {
        return this.litersPerDay;
    }

    /**
     * Returns a new DailyWasteDemandLitersPerDay with the updated value.
     *
     * @param newLitersPerDay New demand value in liters per day
     * @return New DailyWasteDemandLitersPerDay
     * @throws IllegalArgumentException if newLitersPerDay is negative
     */
    public DailyWasteDemandLitersPerDay setLitersPerDay(double newLitersPerDay) {
        return new DailyWasteDemandLitersPerDay(newLitersPerDay);
    }

    /**
     * Adds two DailyWasteDemandLitersPerDay instances.
     *
     * @param other Other DailyWasteDemandLitersPerDay
     * @return New DailyWasteDemandLitersPerDay with the sum
     * @throws IllegalArgumentException if the other object is null
     */
    public DailyWasteDemandLitersPerDay add(DailyWasteDemandLitersPerDay other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_DEMAND_NULL);
        }
        return new DailyWasteDemandLitersPerDay(this.litersPerDay + other.litersPerDay);
    }

    /**
     * Compares if this demand is greater than another.
     *
     * @param other Other DailyWasteDemandLitersPerDay
     * @return true if greater, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThan(DailyWasteDemandLitersPerDay other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_DEMAND_NULL);
        }
        return this.litersPerDay > other.litersPerDay;
    }

    /**
     * Compares if this demand is greater than or equal to another.
     *
     * @param other Other DailyWasteDemandLitersPerDay
     * @return true if greater or equal, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThanOrEqual(DailyWasteDemandLitersPerDay other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_DEMAND_NULL);
        }
        return this.litersPerDay >= other.litersPerDay;
    }

    /**
     * Compares if this demand is less than another.
     *
     * @param other Other DailyWasteDemandLitersPerDay
     * @return true if less, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean lessThan(DailyWasteDemandLitersPerDay other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_DEMAND_NULL);
        }
        return this.litersPerDay < other.litersPerDay;
    }

    /**
     * Compares equality of two DailyWasteDemandLitersPerDay instances.
     *
     * @param otherObject Other object to compare
     * @return true if equal, false otherwise
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        DailyWasteDemandLitersPerDay other = (DailyWasteDemandLitersPerDay) otherObject;
        return Double.compare(this.litersPerDay, other.litersPerDay) == ZERO;
    }

    /**
     * Returns the hash code of the DailyWasteDemandLitersPerDay.
     *
     * @return hashCode
     */
    @Override
    public int hashCode() {
        return Objects.hash(litersPerDay);
    }

    /**
     * Returns the string representation of the DailyWasteDemandLitersPerDay.
     *
     * @return String
     */
    @Override
    public String toString() {
        return String.format("DailyWasteDemandLitersPerDay={litersPerDay=%s}", this.litersPerDay);
    }
}
